linalg/blockoperator.hpp

```diff
@@ -14,12 +14,36 @@
 
 #include "../config/config.hpp"
 #include "../general/array.hpp"
+#include "tblockoperator.hpp"
 #include "operator.hpp"
 #include "blockvector.hpp"
 
 namespace mfem
 {
 
+//! @class BlockConstOperator
+/**
+ * \brief A class to handle constant Block systems in a matrix-free implementation.
+ *
+ * Usage:
+ * - Use one of the constructors to define the block structure.
+ * - Use SetDiagonalBlock or SetBlock to fill the BlockConstOperator
+ * - Use the method Mult and MultTranspose to apply the operator to a vector.
+ *
+ * If a block is not set, it is assumed to be a zero block.
+ */
```

Comment:
Given the discussion in the meeting earlier, I might suggest using this pattern:
https://godbolt.org/z/a68eqh7ET

The compiler can then figure the cv status of the bindings and invalidate the methods as needed. To avoid everything needing to be a header, can then keep the method bodies in a cpp, and use explicit instantiation for `Operator` and `const Operator` template arguments. I think that would work with the type aliasing, but you'd need to check.

Reply:
Yes, I am going to change the classes to typedefs or usings 😉 I was just trapped in some meetings...
Instantiations do not make much sense, I believe, because we want the generality of templates like `TBlockOperator<HyperParMatrix>`.

Reply:
Or do you mean just for compilation to get them to the `blockoperator.o` module?

Reply:
I was working under the assumption it's a drop in to allow `const` of the underlying type, rather than allowing a full selection of all possible types. I'm generally of the view that templates should be avoided unless absolutely necessary, for either reduction in LOC or measurable performance gains. A specialization for each derived type rather than operating on the base type seems like overkill for a class which is only operating through the pointer to operator. The trading away of the C cast for an extra `<>` everywhere I think is quite undesirable, especially when `dynamic_cast` would work fine.

Additionally, you'll be opening up some horrible possibilities for function argument specialization: e.g.
`void foo(const TBlockOperator<A> &t)` wouldn't work for an instance of `BlockOperator<B>` passed in even if `B : A` , as the inheritance is on the arguments not the base. This kind of issue also grows combinatorically if allowing conversion from `ConstBlockOperator` to and from `const BlockOperator` (which I think is a necessary feature to this abstraction proposed here). 

My 2 cents basically boils down to this: templates are infectious at a given abstraction level, introduce one and you start having to put them everywhere to avoid these types of issues. Better to reserve their use for very low level optimization, explicit enumeration tricks or full blown metaprogramming in the type system.

Reply:
Well, it would be an overkill if people used the class for its main purpose, which is to apply it as an operator, which would need only const access to the underlying operators (as I originally wanted). Unfortunately, this is not the case as it turned out people use the class as a collection/array of operators. They then static or C cast the returned operator from `GetBlock`. Sure, dynamic casts would solve it, but things like that DPG miniapps automatically construct only `HyperParMatrix`s internally, so it is understandable they resort to static casts (actually C-style casts to make the things worse). Using the template, the classes can directly indicate what they are constructing, so the constructs can be utilized without any casting. However, I do not see many other use cases in MFEM than `HyperParMatrix`. So definitely I do not want to go the way of overspecialization. If a user needs it for something, it is there, but otherwise I would keep `BlockOperator` as the main thing.

Reply:
Regarding conversion, we have discussed it above with @v-dobrev  👆 . The class does not have an assignment operator, so an additional converting one then does not make sense. I may add a converting copy constructor if that is desired...

Reply:
Ohh, I did not realize, the copy default copy constructor is also deleted because of the deleted assignment, so it does not make sense too. Perhaps the question is why they are deleted? I do not see any reason why it could not do a shallow copy... 🤔

Reply:
> Well, it would be an overkill if people used the class for its main purpose, which is to apply it as an operator, which would need only const access to the underlying operators (as I originally wanted). Unfortunately, this is not the case as it turned out people use the class as a collection/array of operators. They then static or C cast the returned operator from `GetBlock`. Sure, dynamic casts would solve it, but things like that DPG miniapps automatically construct only `HyperParMatrix`s internally, so it is understandable they resort to static casts (actually C-style casts to make the things worse). Using the template, the class methods can directly indicate what they are constructing, so the constructs can be utilized without any casting. However, I do not see many other use cases in MFEM than `HyperParMatrix`. So definitely I do not want to go the way of overspecialization. If a user needs it for something, it is there, but otherwise I would keep `BlockOperator` as the main thing.

@najlkin , I think this is starting to become a bigger issue for no obvious reason. I think things were much simpler the way they were before (having just the BlockOperator return both const and non const version of a block). It's not uncommon to be needing a non-const return of a block, since these blocks are often used to construct preconditioners. There are also other use cases where the blocks aren't even the same type (e.g you might have a block that is a HypreParMatrix or a SparseMatrix or even a "MatrixFreeOperator") which would have no benefit from this change. We can also change the C-style casting to dynamic cast in the DPG miniapps with no issue.

Reply:
I have not decided on a preference for this implementation vs. the old one, but if we revert back to the old `BlockOperator` I would suggest adding a `GetBlockAs` template method (similar to those which `OperatorHandle` has, in order to handle the choice to dynamic or static (when you know) cast:

```
   /** @brief Return the Operator pointer statically cast to a specified OpType.
       Similar to the method Get(). */
   template <typename OpType>
   OpType *As() const { return static_cast<OpType*>(oper); }

   /// Return the Operator pointer dynamically cast to a specified OpType.
   template <typename OpType>
   OpType *Is() const { return dynamic_cast<OpType*>(oper); }
```

That solves the dynamic/static/C-style casting concern in user code at least.

Reply:
@psocratis , I originally kept the things simple if you look to the history, but @v-dobrev was proposing that template version, supported by @tzanio , and I was also considering that as a possible solution, so I have changed that to it. I see only benefits from that, as you have the old `BlockOperator` and on top of that the const and generalized version, which can be used to clean the code when some methods construct only something specific like that `HypreParMatrix`. However, it does _not_ mean you have to use it and I also do not expect to be much used with anything else than `HypreParMatrix`.

Anyway, I agree that the discussion is becoming too wide with that conversion when `BlockOperator` does not have even a copy constructor. Adding it goes beyond this PR and (non-)constness can be included within construction from convertible data types (like in `Array` for example).